src/pykiso/lib/auxiliaries/ykush_auxiliary.py

```diff
@@ -0,0 +1,406 @@
+##########################################################################
+
+"""
+Ykush Auxiliary
+***************
+
+:module: ykush_auxiliary
+
+:synopsis: Auxiliary that can power on and off ports on an Ykush device.
+
+.. currentmodule:: ykush_auxiliary
+
+"""
+import logging
+from contextlib import contextmanager
+from typing import Any, List, Optional, Tuple
+
+import hid
+
+from pykiso.interfaces.dt_auxiliary import DTAuxiliaryInterface
+
+log = logging.getLogger(__name__)
+
+# YKUSH device USB VID
+YKUSH_USB_VID = 0x04D8
+# YKUSH PIDs when in normal operation mode: YKUSH beta, YKUSH, YKUSH3, YKUSHXS
+YKUSH_USB_PID_LIST = (0x0042, 0xF2F7, 0xF11B, 0xF0CD)
+
+# YKUSH device USB comm declarations
+YKUSH_USB_TIMEOUT = 1000  # timeout in ms
+YKUSH_USB_PACKET_SIZE = 64
+YKUSH_USB_PACKET_PAYLOAD_SIZE = 20
+
+# YKUSH device protocol status declarations
+YKUSH_PROTO_OK_STATUS = 1
+
+# YKUSH port state meaning declarations
+YKUSH_PORT_STATE_ON = 1
+YKUSH_PORT_STATE_OFF = 0
+YKUSH_PORT_STATE_ERROR = 255
+YKUSH_PORT_STATE_DICT = {0: "OFF", 1: "ON", 255: "ERROR"}
+
+
+class YkushError(Exception):
+    """General Ykush specific exception used as basis for all others."""
+
+    pass
+
+
+class YkushDeviceNotFound(YkushError):
+    """Raised when no Ykush device is found."""
+
+    pass
+
+
+class YkushStatePortNotRetrieved(YkushError):
+    """Raised when the state of a port can't be retrieved."""
+
+    pass
+
+
+class YkushSetStateError(YkushError):
+    """Raised when a port couldn't be switched on or off."""
+
+    pass
+
+
+class YkushPortNumberError(YkushError):
+    """Raised when the port number doesn't exist."""
+
+    pass
+
+
+class YkushAuxiliary(DTAuxiliaryInterface):
+    """Auxiliary used to power on and off the ports."""
+
+    def __init__(self, serial_number: str = None, **kwargs):
+        """Initialize attribute
+
+        :param serial: Serial number of the device to connect, if he is not defined
+            then it will connect to the first Ykush device it find, defaults to None.
+        :raises YkushDeviceNotFound: If no device is found or the serial number
+            is not the serial of one device.
+        """
+        super().__init__(
+            is_proxy_capable=False,
+            tx_task_on=False,
+            rx_task_on=False,
+            connector_required=False,
+            **kwargs,
+        )
+        self._ykush_device = None
+        self._product_id = None
+        self._path = None
+        self.find_device(serial_number)
+        self.number_of_port = self.get_number_of_port()
+
+    def _create_auxiliary_instance(self) -> bool:
+        """Power on all port at the start.
+        :return: True if succesful
+        """
+        log.internal_info("Create auxiliary instance")
+        self.set_all_ports_on()
+        return True
+
+    def _delete_auxiliary_instance(self) -> bool:
+        """Power on all port to restore the ports.
+        :return: always True
+        """
+        self.set_all_ports_on()
```

Comment:
Should the auxiliary switch ports implicitly?

Reply:
I thought if for example you do a test with the ykush auxiliary where you put a port off and it stays off at the end of the test. And after you want to do other tests without the auxiliary then you will not be able to communicate with the device on the port off. Which is why I reset port at the end of the test.